File: engine/solver.js
/**
 * STEP 4 — Basic Solver (whole-number, non-negative)
 *
 * Solves a normalized LINEAR equation:
 *   a1*x1 + a2*x2 + ... + an*xn = constant
 *
 * Requirements for this step:
 * - whole-number solutions only
 * - non-negative values only
 * - use backtracking (not nested brute-force loops)
 *
 * NOTE (intentional for Step 4):
 * - This solver currently supports POSITIVE integer coefficients and a non-negative constant.
 * - Edge cases like negative/zero coefficients, infinite solutions, etc. are handled in Step 6.
 */

/** @typedef {{ coefficients: Record<string, number>, constant: number }} NormalizedEquation */

const {
  normalizeConstraints,
  domainFor,
  satisfiesAll,
} = require("./constraints");

function assertSafeInteger(n, label) {
  if (!Number.isSafeInteger(n)) throw new Error(`${label} must be a safe integer.`);
}

function validateNormalizedInput(eq) {
  if (!eq || typeof eq !== "object") throw new Error("Normalized equation is required.");
  if (!eq.coefficients || typeof eq.coefficients !== "object") {
    throw new Error("coefficients must be an object.");
  }

  assertSafeInteger(eq.constant, "constant");
  if (eq.constant < 0) {
    throw new Error("Step 4 solver expects a non-negative constant.");
  }

  const vars = Object.keys(eq.coefficients);
  if (vars.length === 0) throw new Error("No variables found to solve for.");

  for (const v of vars) {
    const a = eq.coefficients[v];
    assertSafeInteger(a, `coefficient for ${v}`);
    if (a <= 0) {
      throw new Error(
        `Step 4 solver expects positive coefficients only; got ${v}: ${a}`
      );
    }
  }

  return vars;
}

function cloneSolution(solution) {
  return { ...solution };
}

function gcd(a, b) {
  a = Math.abs(a);
  b = Math.abs(b);
  while (b !== 0) {
    const t = a % b;
    a = b;
    b = t;
  }
  return a;
}

function gcdAll(values) {
  let g = 0;
  for (const v of values) g = gcd(g, v);
  return g;
}

/**
 * Solve 1-variable equation a*x = c.
 * @param {string} variable
 * @param {number} a
 * @param {number} c
 */
function solve1(variable, a, c) {
  if (c % a !== 0) return [];
  const x = c / a;
  if (x < 0) return [];
  return [{ [variable]: x }];
}

/**
 * General N-variable backtracking solver.
 * Deterministic order: variables are sorted alphabetically.
 * @param {NormalizedEquation} eq
 * @param {Record<string, any> | undefined | null} [constraintsRaw]
 * @param {{ maxResults?: number }} [options]
 * @returns {Array<Record<string, number>>}
 */
function solveNormalized(eq, constraintsRaw, options) {
  const maxResults = options?.maxResults ?? 200;
  const vars = validateNormalizedInput(eq);
  const coeffs = eq.coefficients;
  const constant = eq.constant;

   const constraints = normalizeConstraints(constraintsRaw);

  // Variable ordering heuristic: larger coefficient first => smaller branching factor.
  // Deterministic: tie-break by variable name.
  const variables = vars
    .slice()
    .sort((a, b) => (coeffs[b] - coeffs[a]) || a.localeCompare(b));

  if (variables.length === 1) {
    const v = variables[0];
    const base = solve1(v, coeffs[v], constant);
    if (!base.length) return [];

    const x = base[0][v];
    const dom = domainFor(v, constraints, x);
    if (x < dom.min || x > dom.max) return [];
    if (!satisfiesAll(base[0], constraints)) return [];
    return base;
  }

  /** @type {Array<Record<string, number>>} */
  const solutions = [];

  function reachedLimit() {
    return solutions.length >= maxResults;
  }

  /**
   * Prune using constraint-aware min/max bounds and gcd divisibility.
   * @param {number} index
   * @param {number} remaining
   */
  function canStillSolve(index, remaining) {
    // Compute shifted remaining after applying mins for variables index..end.
    let minSum = 0;
    let maxSum = 0;

    /** @type {number[]} */
    const aList = [];

    for (let j = index; j < variables.length; j++) {
      const v = variables[j];
      const a = coeffs[v];
      aList.push(a);

      const hardMax = Math.floor(remaining / a);
      const dom = domainFor(v, constraints, hardMax);

      if (dom.max < dom.min) return false; // empty domain

      minSum += a * dom.min;
      maxSum += a * dom.max;
      if (minSum > remaining) return false;
    }

    if (remaining < minSum) return false;
    if (remaining > maxSum) return false;

    const shifted = remaining - minSum;
    const g = gcdAll(aList);
    if (g !== 0 && shifted % g !== 0) return false;

    return true;
  }

  /**
   * @param {number} index
   * @param {number} remaining
   * @param {Record<string, number>} partial
   */
  function backtrack(index, remaining, partial) {
    if (reachedLimit()) return;
    if (!canStillSolve(index, remaining)) return;

    const v = variables[index];
    const a = coeffs[v];

    if (index === variables.length - 1) {
      // Last variable: direct compute.
      if (remaining % a !== 0) return;
      const x = remaining / a;
      if (x < 0) return;
      const candidate = { ...partial, [v]: x };
      if (satisfiesAll(candidate, constraints)) {
        solutions.push(candidate);
      }
      return;
    }

    const hardMax = Math.floor(remaining / a);
    const dom = domainFor(v, constraints, hardMax);
    for (let x = dom.min; x <= dom.max; x += dom.step) {
      const nextRemaining = remaining - a * x;
      // nextRemaining is always >= 0 due to max
      partial[v] = x;
      backtrack(index + 1, nextRemaining, partial);

      if (reachedLimit()) break;
    }

    delete partial[v];
  }

  backtrack(0, constant, {});
  return solutions;
}

module.exports = {
  solveNormalized,
};

File: engine/rust/src/solver.rs
// ============================================================
//  solver.rs — Port of engine/solver.js
//
//  Backtracking solver for a normalized linear equation:
//    a1*x1 + a2*x2 + ... = constant
//
//  - Non-negative whole-number solutions only
//  - Constraint-aware domain pruning
//  - GCD-based divisibility pruning (key performance win)
//  - Variables ordered by descending coefficient (smaller branching)
// ============================================================

use std::collections::HashMap;
use crate::normalizer::LinearForm;
use crate::constraints::{Constraints, domain_for, satisfies_all};

fn gcd(a: i64, b: i64) -> i64 {
    let (mut a, mut b) = (a.abs(), b.abs());
    while b != 0 {
        let t = a % b;
        a = b;
        b = t;
    }
    a
}

fn gcd_all(values: &[i64]) -> i64 {
    values.iter().fold(0, |g, &v| gcd(g, v))
}

/// Solve a 1-variable sub-problem: a*x = c
fn solve_one(var: char, a: i64, c: i64) -> Option<HashMap<char, i64>> {
    if c % a != 0 { return None; }
    let x = c / a;
    if x < 0 { return None; }
    let mut m = HashMap::new();
    m.insert(var, x);
    Some(m)
}

pub struct SolverOptions {
    pub max_results: usize,
}

/// Solve the normalized equation with optional constraints.
/// Returns up to `max_results` solutions.
pub fn solve_normalized(
    eq: &LinearForm,
    constraints: &Constraints,
    options: &SolverOptions,
) -> Vec<HashMap<char, i64>> {
    let max_results = options.max_results;

    // Collect and sort variables: larger coefficient first (smaller branching factor).
    // Tie-break alphabetically for deterministic output.
    let mut variables: Vec<char> = eq.coefficients.keys().cloned().collect();
    variables.sort_by(|&a, &b| {
        let ca = eq.coefficients[&a];
        let cb = eq.coefficients[&b];
        cb.cmp(&ca).then(a.cmp(&b))
    });

    if variables.is_empty() {
        return vec![];
    }

    let coeffs = &eq.coefficients;
    let constant = eq.constant;

    // 1-variable fast path
    if variables.len() == 1 {
        let v = variables[0];
        let a = coeffs[&v];
        match solve_one(v, a, constant) {
            None => return vec![],
            Some(sol) => {
                let dom = domain_for(v, constraints, sol[&v]);
                if sol[&v] < dom.min || sol[&v] > dom.max { return vec![]; }
                if !satisfies_all(&sol, constraints) { return vec![]; }
                return vec![sol];
            }
        }
    }

    let mut solutions: Vec<HashMap<char, i64>> = Vec::new();
    let mut partial: HashMap<char, i64> = HashMap::new();

    backtrack(
        0,
        constant,
        &variables,
        coeffs,
        constraints,
        max_results,
        &mut partial,
        &mut solutions,
    );

    solutions
}

#[allow(clippy::too_many_arguments)]
fn backtrack(
    index: usize,
    remaining: i64,
    variables: &[char],
    coeffs: &HashMap<char, i64>,
    constraints: &Constraints,
    max_results: usize,
    partial: &mut HashMap<char, i64>,
    solutions: &mut Vec<HashMap<char, i64>>,
) {
    if solutions.len() >= max_results { return; }
    if !can_still_solve(index, remaining, variables, coeffs, constraints) { return; }

    let v = variables[index];
    let a = coeffs[&v];

    if index == variables.len() - 1 {
        // Last variable: direct compute
        if remaining % a != 0 { return; }
        let x = remaining / a;
        if x < 0 { return; }
        let mut candidate = partial.clone();
        candidate.insert(v, x);
        if satisfies_all(&candidate, constraints) {
            solutions.push(candidate);
        }
        return;
    }

    let hard_max = remaining / a;   // floor division (both non-negative here)
    let dom = domain_for(v, constraints, hard_max);

    let mut x = dom.min;
    while x <= dom.max {
        if solutions.len() >= max_results { break; }
        let next_remaining = remaining - a * x;
        partial.insert(v, x);
        backtrack(index + 1, next_remaining, variables, coeffs, constraints, max_results, partial, solutions);
        x += dom.step;
    }

    partial.remove(&v);
}

/// Prune: can variables[index..] still sum to `remaining`?
///
/// Checks:
///  1. min-sum > remaining → impossible
///  2. max-sum < remaining → impossible
///  3. (remaining - min-sum) % gcd(coeffs) ≠ 0 → impossible
fn can_still_solve(
    index: usize,
    remaining: i64,
    variables: &[char],
    coeffs: &HashMap<char, i64>,
    constraints: &Constraints,
) -> bool {
    if remaining < 0 { return false; }

    let mut min_sum = 0i64;
    let mut max_sum = 0i64;
    let mut a_list: Vec<i64> = Vec::new();

    for j in index..variables.len() {
        let v = variables[j];
        let a = coeffs[&v];
        a_list.push(a);

        let hard_max = remaining / a;  // approximate upper bound
        let dom = domain_for(v, constraints, hard_max);

        if dom.is_empty() { return false; }

        min_sum += a * dom.min;
        max_sum += a * dom.max;

        if min_sum > remaining { return false; }
    }

    if remaining < min_sum { return false; }
    if remaining > max_sum { return false; }

    let shifted = remaining - min_sum;
    let g = gcd_all(&a_list);
    if g != 0 && shifted % g != 0 { return false; }

    true
}
